draft-ietf-quic-transport.md

```diff
@@ -2497,16 +2496,32 @@ handshake is confirmed (see Section 4.1.2 of {{QUIC-TLS}}), an endpoint MUST
 send any CONNECTION_CLOSE frames in a 1-RTT packet.  However, prior to
 confirming the handshake, it is possible that more advanced packet protection
 keys are not available to the peer, so the frame MAY be replicated in a packet
-that uses a lower packet protection level.
+that uses a lower packet protection level.  More specifically:
+
+* A client will always know whether the server has Handshake keys (see
+  {{discard-initial}}), but it is possible that a server does not know whether
+  the client has Handshake keys.  Under these circumstances, a server SHOULD
+  send a CONNECTION_CLOSE frame in both Handshake and Initial packets to ensure
+  that at least one of them is processable by the client.
+
+* A client that sends CONNECTION_CLOSE in a 0-RTT packet cannot be assured of
+  the server has accepted 0-RTT and so sending a CONNECTION_CLOSE frame in an
+  Initial packet makes it more likely that the server can receive the close
+  signal, even if the application error code might not be received.
+
+* Prior to confirming the handshake, a peer might be unable to process 1-RTT
+  packets, so an endpoint SHOULD send CONNECTION_CLOSE in both Handshake and
+  1-RTT packets.  A server SHOULD also send CONNECTION_CLOSE in an Initial
+  packet.
+
+An CONNECTION_CLOSE of type 0x1d MUST be replaced by a CONNECTION_CLOSE of type
+0x1c when sending the frame in Initial packets. Otherwise, information about
+the application state might be revealed. Endpoints MUST clear the value of the
+Reason Phrase field and SHOULD use the APPLICATION_ERROR code when converting
+to a CONNECTION_CLOSE of type 0x1c.
```

Comment:
```suggestion
Sending a CONNECTION_CLOSE of type 0x1d in an Initial packet might reveal application
state to third parties. To avoid this leakage, a CONNECTION_CLOSE of type 0x1d MUST
be replaced by a CONNECTION_CLOSE of type 0x1c when sending the frame in Initial
packets. Endpoints MUST clear the value of the Reason Phrase field and SHOULD use
the APPLICATION_ERROR code when converting to a CONNECTION_CLOSE of type 0x1c.
```